### PR TITLE
fix Empty string convert to DecimalOrNumbeOrDateOrDateTime

### PR DESCRIPTION
**Pull request by sunny19930321**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix (user-visible misbehaviour in official stable or prestable release)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
fix Empty string convert to Decimal Or Numbe Or Date Or DateTime

Detailed description / Documentation draft:
Closes:https://github.com/ClickHouse/ClickHouse/issues/27120

> By adding documentation, you'll allow users to try your new feature immediately, not when someone else will have time to document it later. Documentation is necessary for all features that affect user experience in any way. You can add brief documentation draft above, or add documentation right into your patch as Markdown files in [docs](https://github.com/ClickHouse/ClickHouse/tree/master/docs) folder.

> If you are doing this for the first time, it's recommended to read the lightweight [Contributing to ClickHouse Documentation](https://github.com/ClickHouse/ClickHouse/tree/master/docs/README.md) guide first.

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

File: tests/queries/0_stateless/02013_emptystring_cast.sql [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+drop table if exists test_date;
+create table test_date (`data` Date) engine = MergeTree order by tuple();
+insert into test_date values ('2021-01-01'), (NULL), ('2021-02-01'), ('2021-03-01');
+drop table if exists test_date;
+
+drop table if exists test_datetime;
+create table test_datetime (`data` DateTime) engine = MergeTree order by tuple();
+insert into test_datetime values ('2021-01-01'), (NULL), ('2021-02-01'), ('2021-03-01');
+drop table if exists test_datetime;
```
Comment (kssenii) [inline tests/queries/0_stateless/02013_emptystring_cast.sql:19]: The test with `test_date` actually works without any exception on current mater. But test with `test_datetime` fails. May be it means that the fix should be different. Probably some case is missing in parseImpl?

Comment (sunny19930321) [reply tests/queries/0_stateless/02013_emptystring_cast.sql:19]: Ok, I'm looking at the test questions

**Comment by tavplubix** [comment]
The fix is too ad-hoc and it changes behavior of CAST function.
A better way to fix it is to use more complex expression to convert nullable value to default here:
https://github.com/ClickHouse/ClickHouse/blob/9ca422f0c5380124b57270ad18505e572d1b3afb/src/Processors/Formats/Impl/ConstantExpressionTemplate.cpp#L635-L640

We can use something like `CAST(ifNull(x AS _expression, CAST(defaultValueOfTypeName('TypeName'), toTypeName(assumeNotNull(_expression)))), 'TypeName')` instead of just `CAST(assumeNotNull(expr as _expression), 'TypeName')`.
Explanation:
 - `defaultValueOfTypeName('TypeName')` is default value of desired type
 - `toTypeName(assumeNotNull(_expression))` is name of type of expression without `Nullable`
 - `CAST(defaultValueOfTypeName('TypeName'), toTypeName(assumeNotNull(_expression)))), 'TypeName')` is the value of expression type that will be casted to default value of desired type
 - `ifNull(x AS _expression, CAST(defaultValueOfTypeName('TypeName'), toTypeName(assumeNotNull(_expression)))` is value of expression or value that will be casted to default
 
Another option is to use short circuit evaluation (see #23367). The fix probably will be simpler, but it will not be possible to backport it.

**Comment by sunny19930321** [comment]
> The fix is too ad-hoc and it changes behavior of CAST function.
> A better way to fix it is to use more complex expression to convert nullable value to default here:
> https://github.com/ClickHouse/ClickHouse/blob/9ca422f0c5380124b57270ad18505e572d1b3afb/src/Processors/Formats/Impl/ConstantExpressionTemplate.cpp#L635-L640
> 
> We can use something like `CAST(ifNull(x AS _expression, CAST(defaultValueOfTypeName('TypeName'), toTypeName(assumeNotNull(_expression)))), 'TypeName')` instead of just `CAST(assumeNotNull(expr as _expression), 'TypeName')`.
> Explanation:
> 
> * `defaultValueOfTypeName('TypeName')` is default value of desired type
> * `toTypeName(assumeNotNull(_expression))` is name of type of expression without `Nullable`
> * `CAST(defaultValueOfTypeName('TypeName'), toTypeName(assumeNotNull(_expression)))), 'TypeName')` is the value of expression type that will be casted to default value of desired type
> * `ifNull(x AS _expression, CAST(defaultValueOfTypeName('TypeName'), toTypeName(assumeNotNull(_expression)))` is value of expression or value that will be casted to default
> 
> Another option is to use short circuit evaluation (see #23367). The fix probably will be simpler, but it will not be possible to backport it.



> The fix is too ad-hoc and it changes behavior of CAST function.
> A better way to fix it is to use more complex expression to convert nullable value to default here:
> https://github.com/ClickHouse/ClickHouse/blob/9ca422f0c5380124b57270ad18505e572d1b3afb/src/Processors/Formats/Impl/ConstantExpressionTemplate.cpp#L635-L640
> 
> We can use something like `CAST(ifNull(x AS _expression, CAST(defaultValueOfTypeName('TypeName'), toTypeName(assumeNotNull(_expression)))), 'TypeName')` instead of just `CAST(assumeNotNull(expr as _expression), 'TypeName')`.
> Explanation:
> 
> * `defaultValueOfTypeName('TypeName')` is default value of desired type
> * `toTypeName(assumeNotNull(_expression))` is name of type of expression without `Nullable`
> * `CAST(defaultValueOfTypeName('TypeName'), toTypeName(assumeNotNull(_expression)))), 'TypeName')` is the value of expression type that will be casted to default value of desired type
> * `ifNull(x AS _expression, CAST(defaultValueOfTypeName('TypeName'), toTypeName(assumeNotNull(_expression)))` is value of expression or value that will be casted to default
> 
> Another option is to use short circuit evaluation (see #23367). The fix probably will be simpler, but it will not be possible to backport it.

ok, thanks

**Comment by sunny19930321** [comment]
thanks for your advice

**Comment by tavplubix** [comment]
Merged in #28413

**Pull request closed without merging**